server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -851,6 +852,12 @@ private void innerExecute(
                 itemDroppedHandler.accept(slot);
                 handler.accept(null);
             } else {
+                try {
+                    CollectionUtils.ensureNoSelfReferences(result.getSourceAndMetadata(), "ingest pipeline [" + pipeline.getId() + "]");
```

Comment:
Is this checked for each pipeline? I think we should only execute this for pipelines that have scripts in it?

Reply:
As as far as I know, there is no easy of checking whether a pipeline uses scripts.
So I think adding this logic is currently not worth it. If at some point in time, 
there are use cases where with pipelines with no scripts that have a performance a reak regression,
then we can think about adding this logic, so that we self reference check can be skipped.

Reply:
So it turns out scripts in conditionals can't change the document (see `ConditionalProcessor#evaluate(...)`).
We can use `ScriptService#getProcessorsInPipeline(...)` to find script processors, but the issue here is that `ScriptProcessor` is defined in a ingest-common module. But maybe we can get processors by type?

Reply:
To clarify, scripts in conditions don't need to be checked for self referencing. Only script processors should be checked. I think we can add a method to `ScriptService` that fetches pipeline with processors of a specific type, just like `getProcessorsInPipeline(...)` does for fetching pipelines with processor of a certain `Class`.

Reply:
We talked about this and this optimization can be added in a followup pr.

Reply:
Open #87335 for this followup, so we don't forget.